spec/src/main/asciidoc/query-language.asciidoc

```diff
@@ -0,0 +1,341 @@
+- A `not` operator expression is never satisfied if its operand _is_ satisfied.
+
+This specification leaves undefined the interpretation of the `not` operator when its operand _is not_ satisfied.
+
+CAUTION: A compliant implementation of JDQL might feature SQL/JPQL-style ternary logic, where `not n > 0` is an unsatisfied logical expression when `n` evaluates to null, or it might feature binary logic where the same expression is considered satisfied. Application programmers should take great care when using the `not` operator with scalar expressions involving `null` values.
+
+=== Clauses
+
+Each JDQL statement is built from a sequence of _clauses_. The beginning of a clause is identified by a keyword: `from`, `where`, `select`, `set`, or `order`.
+
+There is a logical ordering of clauses, reflecting the order in which their effect must be computed by the datastore:
+
+1. `from`
+2. `where`,
+3. `select` or `set`,
+4. `order`.
+
+The interpretation and effect of each clause in this list is influenced by clauses occurring earlier in the list, but not by clauses occurring later in the list.
+
+==== From clause
+
+The `from` clause, with syntax given by `from_clause`, specifies an _entity name_ which identifies the queried entity. Path expressions occurring in later clauses are interpreted with respect to this entity. That is, the first element of each path expression in the query must be a persistent field of the entity named in the `from` clause. The entity name is a Java identifier, usually the unqualified name of the entity class, as specified in <<Entity Names>>.
+
+The `from` clause is optional in `select` statements. When it is missing, the queried entity is determined by the return type of the repository method, or, if the return type is not an entity type, by the primary entity type of the repository.
+
+NOTE: The syntax of the `update` statement is irregular, with the `from` keyword implied. That is, the syntax _should_ be `update from Entity`, but for historical reasons it is simply `update Entity`.
+
+==== Where clause
+
+The `where` clause, with syntax given by `where_clause`, specifies a conditional expression used to restrict the records returned, deleted, or updated by the query. Only records for which the conditional expression is satisfied are returned, deleted, or updated.
+
+The `where` clause is always optional. When it is missing, there is no restriction, and, and all records are of the queried entity type are returned, deleted, or updated.
+
+==== Select clause
+
+The `select` clause, with syntax given by `select_clause`, specifies a list of expressions which are returned by the query. Each expression in the list is evaluated for each record which satisfies the restriction imposed by the `where` clause. Alternatively, the `select` clause may contain a single `count(*)` aggregate expression, which evaluates to the number of records which satisfy the restriction.
```

Comment:
what happens in the db doesn't support `count(*)`?

Reply:
I guess we put the Unsupportedoperationexception, am I right, @njr-11 ?

Reply:
@graemerocher @otaviojava please review d127e7731db03a209101278ace53c2cf13933d41

Reply:
> I guess we put the Unsupportedoperationexception, am I right, @njr-11 ?

Right, based on how we treat other unsupported function in the spec, it should be `UnsupportedOperationException`.  I think the spec is actually missing this at the level of `countBy`. It specifies using the exception for keywords used in the remained of Query by Method Name that aren't supported but doesn't currently say that method name prefixes like `count` can be unsupported.